### Exploring Ebay Car Sales Data

In this project, we clean and analyze a dataset of used car listings on a classifieds section of the German eBay website, eBay Kleinazeigen. We will use pandas to make this process quick and easy.

The orignal dataset is available from Kaggle. While the dataset we use is a sample of 50,000 data points prepared and made messier by Dataquest.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding="Latin-1")

In [2]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04            GOLF_4_1_4__3T�RER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

    price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480$   test         NaN                1993    manuell        0   golf   
1  18300$   test       coupe                2011    manuell      190    NaN   
2   9800$   test         suv                2004  automatik      163  grand   
3   1500$   test  kleinwagen                2001    manuell       75   golf   
4   3600$   test  kleinwagen                2008    manuell       69  fabia   

  kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0  150000km                    0   benzin  volkswagen               NaN   
1  125000km                    5   diesel        audi                ja   
2  125000km                    8   diesel        jeep               NaN   
3  150000km                    6   benzin  volkswagen              nein   
4   90000km                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49996 entries, 0 to 49995
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          49996 non-null  object
 1   name                 49996 non-null  object
 2   seller               49996 non-null  object
 3   offerType            49996 non-null  object
 4   price                49996 non-null  object
 5   abtest               49996 non-null  object
 6   vehicleType          44832 non-null  object
 7   yearOfRegistration   49996 non-null  int64 
 8   gearbox              47266 non-null  object
 9   powerPS              49996 non-null  int64 
 10  model                47241 non-null  object
 11  kilometer            49996 non-null  object
 12  monthOfRegistration  49996 non-null  int64 
 13  fuelType             45387 non-null  object
 14  brand                49996 non-null  object
 15  notRepairedDamage    40270 non-null  object
 16  date

We have 20 columsn in the dataset, of which 15 have no nulls. We have a few cells that can be converted to numeric including price and odometer. The notRepairedDamage column can be converted to boolean and the column names could be cleaned up a bit.

### Cleaning Column Names

First we convert the colums names from camelcase to snakecase which is preferred in Python.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_ps', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04            GOLF_4_1_4__3T�RER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

    price abtest vehicle_type  year_of_registration    gearbox  power_ps  \
0    480$   test          NaN                  1993    manuell         0   
1  18300$   test        coupe                  2011    manuell       190   
2   9800$   test          suv                  2004  automatik       163   
3   1500$   test   kleinwagen                  2001    manuell        75   
4   3600$   test   kleinwagen                  2008    manuell        69   

   model  odometer  month_of_registration fuel_type       brand  \
0   golf  150000km                      0    benzin  volkswagen   
1    NaN  125000km                      5    diesel        audi   
2  grand  125000km                      8    diesel        jeep   
3   golf  150000km                      6    benzin  volkswagen   
4  fabia   90000km                      7    diesel       skoda   

  not_repaired_damage         date_created  nr_of_pictures  postal_code  \
0                 NaN  2016-03-24 00:00:00               0        70435   
1                  ja  2016-03-24 00:00:00               0        66954   
2                 NaN  2016-03-14 00:00:00               0        90480   
3                nein  2016-03-17 00:00:00               0        91074   
4                nein  2016-03-31 00:00:00               0        60437   

             last_seen  
0  2016-04-07 03:16:57  
1  2016-04-07 01:46:50  
2  2016-04-05 12:47:46  
3  2016-03-17 17:40:17  
4  2016-04-06 10:17:21

### Initial Exploring and Cleaning

The price column has values stored as strings and the $ is at the end of the string.

In [6]:
# eliminate $ and transform string to integer
def price(column):
    string = column
    string = string.replace('$','')
    return int(string)
    
autos['price'] = autos['price'].apply(price)
autos['price'].head()

0      480
1    18300
2     9800
3     1500
4     3600
Name: price, dtype: int64

In [7]:
# minus values?
autos['price'].describe()

count    4.999600e+04
mean     6.859421e+03
std      1.052015e+05
min     -2.200000e+04
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.234568e+07
Name: price, dtype: float64

The lowest prise is -2200 USD, which is unrealistic, I should multiply this values by -1.

In [8]:
def min_plus(column):
    string = column
    if string < 0:
        string = string * -1
        return string
    else:
        return string
    
autos['price'] = autos['price'].apply(min_plus)
autos['price'].describe()
# the min of the column is 0

count    4.999600e+04
mean     6.860849e+03
std      1.052015e+05
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.234568e+07
Name: price, dtype: float64

The kilometer column string and the last 2 charachters are the unit (km), I will replace them and convert into integer.

In [9]:
# eliminate $ and transform string to integer
def km_remov(column):
    string = column
    string = string.replace('km','')
    return int(string)
    
autos['odometer'] = autos['odometer'].apply(km_remov)
autos['odometer'].head()

0    150000
1    125000
2    125000
3    150000
4     90000
Name: odometer, dtype: int64

In [10]:
autos['odometer'].describe()

count     49996.000000
mean     125507.040563
std       40120.454827
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer, dtype: float64

In [11]:
autos = autos.rename(columns = {'odometer':'odometer_km'})
autos = autos.rename(columns = {'price':'price_$'})

In [12]:
autos.describe(include='all')

date_crawled         name  seller offer_type       price_$  \
count                 49996        49996   49996      49996  4.999600e+04   
unique                48239        38784       1          2           NaN   
top     2016-03-12 15:46:28  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           97   49996      49995           NaN   
mean                    NaN          NaN     NaN        NaN  6.860849e+03   
std                     NaN          NaN     NaN        NaN  1.052015e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  1.234568e+07   

       abtest vehicle_type  year_of_registration  gearbox      power_ps  \
count   49996        44832          49996.000000    47266  49996.000000   
unique      2            8                   NaN        2           NaN   
top      test    limousine                   NaN  manuell           NaN   
freq    25940        12887                   NaN    36964           NaN   
mean      NaN          NaN           2004.683015      NaN    113.663033   
std       NaN          NaN             95.090032      NaN    129.515728   
min       NaN          NaN           1000.000000      NaN      0.000000   
25%       NaN          NaN           1999.000000      NaN     69.000000   
50%       NaN          NaN           2003.000000      NaN    105.000000   
75%       NaN          NaN           2008.000000      NaN    150.000000   
max       NaN          NaN           9999.000000      NaN  10520.000000   

        model    odometer_km  month_of_registration fuel_type       brand  \
count   47241   49996.000000           49996.000000     45387       49996   
unique    246            NaN                    NaN         7          40   
top      golf            NaN                    NaN    benzin  volkswagen   
freq     3906            NaN                    NaN     29875       10474   
mean      NaN  125507.040563               5.740859       NaN         NaN   
std       NaN   40120.454827               3.720408       NaN         NaN   
min       NaN    5000.000000               0.000000       NaN         NaN   
25%       NaN  100000.000000               3.000000       NaN         NaN   
50%       NaN  150000.000000               6.000000       NaN         NaN   
75%       NaN  150000.000000               9.000000       NaN         NaN   
max       NaN  150000.000000              12.000000       NaN         NaN   

       not_repaired_damage         date_created  nr_of_pictures   postal_code  \
count                40270                49996         49996.0  49996.000000   
unique                   2                   69             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35281                 1986             NaN           NaN   
mean                   NaN                  NaN             0.0  50732.576786   
std                    NaN                  NaN             0.0  25874.946210   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30419.000000   
50%                    NaN                  NaN             0.0  49565.000000   
75%                    NaN                  NaN             0.0  71634.000000   
max                    NaN                  NaN             0.0  99996.000000   

                  last_seen  
count                 49996  
unique                39329  
top     2016-04-07 02:15:37  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                  

In [13]:
print(autos['price_$'].describe(),'\n')
print(autos['price_$'].value_counts().head(15))

count    4.999600e+04
mean     6.860849e+03
std      1.052015e+05
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.234568e+07
Name: price_$, dtype: float64 

0       1463
500      799
1500     714
1000     621
1200     598
2500     581
800      560
3500     502
2000     488
750      480
600      478
999      468
650      392
300      390
700      389
Name: price_$, dtype: int64


1463 vehicle have been sold for free it looks like unrealistic, so I'll remove that rows.

In [14]:
boolean_free = autos['price_$'] != 0
autos = autos[boolean_free]
autos['price_$'].value_counts().head()

500     799
1500    714
1000    621
1200    598
2500    581
Name: price_$, dtype: int64

It looks like more realistic.

In [15]:
autos['odometer_km'].value_counts()

150000    31228
125000     5084
100000     2068
90000      1593
80000      1511
70000      1366
60000      1163
50000      1048
40000       864
30000       830
5000        778
20000       731
10000       269
Name: odometer_km, dtype: int64

The odometer column does not contain unrealistic values.

In [16]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48533 entries, 0 to 49995
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   date_crawled           48533 non-null  object
 1   name                   48533 non-null  object
 2   seller                 48533 non-null  object
 3   offer_type             48533 non-null  object
 4   price_$                48533 non-null  int64 
 5   abtest                 48533 non-null  object
 6   vehicle_type           43847 non-null  object
 7   year_of_registration   48533 non-null  int64 
 8   gearbox                46142 non-null  object
 9   power_ps               48533 non-null  int64 
 10  model                  46067 non-null  object
 11  odometer_km            48533 non-null  int64 
 12  month_of_registration  48533 non-null  int64 
 13  fuel_type              44396 non-null  object
 14  brand                  48533 non-null  object
 15  not_repaired_damage

There are a number of columns that represent date values:

date_crawled
last_seen
ad_created
registration_month
registration_year
In the DataFrame date_crawled, last_seen and ad_created are columns that are all represented by strings. These will need to be converted to numeric representation so we can analyse it. The registration_month and registration_year columns are already represented as numeric values.

First let's see how the values in these three string columns are formatted

In [17]:
autos['date_crawled'] = pd.to_datetime(autos['date_crawled'])
autos['date_created'] = pd.to_datetime(autos['date_created'])
autos['last_seen'] = pd.to_datetime(autos['last_seen'])
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48533 entries, 0 to 49995
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   date_crawled           48533 non-null  datetime64[ns]
 1   name                   48533 non-null  object        
 2   seller                 48533 non-null  object        
 3   offer_type             48533 non-null  object        
 4   price_$                48533 non-null  int64         
 5   abtest                 48533 non-null  object        
 6   vehicle_type           43847 non-null  object        
 7   year_of_registration   48533 non-null  int64         
 8   gearbox                46142 non-null  object        
 9   power_ps               48533 non-null  int64         
 10  model                  46067 non-null  object        
 11  odometer_km            48533 non-null  int64         
 12  month_of_registration  48533 non-null  int64         
 13  f

Dates have been converted into proper datatype.

In [18]:
autos.describe()

price_$  year_of_registration      power_ps    odometer_km  \
count  4.853300e+04          48533.000000  48533.000000   48533.000000   
mean   7.067665e+03           2004.681104    114.479035  125465.353471   
std    1.067685e+05             89.317462    121.723570   39910.645941   
min    1.000000e+00           1000.000000      0.000000    5000.000000   
25%    1.250000e+03           1999.000000     71.000000  100000.000000   
50%    3.000000e+03           2004.000000    105.000000  150000.000000   
75%    7.499000e+03           2008.000000    150.000000  150000.000000   
max    1.234568e+07           9999.000000  10520.000000  150000.000000   

       month_of_registration  nr_of_pictures   postal_code  
count           48533.000000         48533.0  48533.000000  
mean                5.803124             0.0  50923.450230  
std                 3.690942             0.0  25847.291205  
min                 0.000000             0.0   1067.000000  
25%                 3.000000             0.0  30459.000000  
50%                 6.000000             0.0  49716.000000  
75%                 9.000000             0.0  71711.000000  
max                12.000000             0.0  99996.000000

The minimum of the year of registration column is unrealistic, I 1ill filter out rows, where the registration year is out of the 1990-2016 interval. 

In [19]:
boolea_reg_date = autos['year_of_registration'].between(1990,2016)
autos = autos[boolea_reg_date]
autos.describe()

price_$  year_of_registration      power_ps    odometer_km  \
count  4.529200e+04          45292.000000  45292.000000   45292.000000   
mean   6.607758e+03           2003.607171    116.713945  125860.748035   
std    8.116396e+04              5.770750    123.484103   39444.087139   
min    1.000000e+00           1990.000000      0.000000    5000.000000   
25%    1.250000e+03           1999.000000     75.000000  125000.000000   
50%    3.100000e+03           2003.000000    109.000000  150000.000000   
75%    7.500000e+03           2008.000000    150.000000  150000.000000   
max    1.234568e+07           2016.000000  10520.000000  150000.000000   

       month_of_registration  nr_of_pictures   postal_code  
count           45292.000000         45292.0  45292.000000  
mean                5.868564             0.0  51045.867107  
std                 3.670122             0.0  25826.805092  
min                 0.000000             0.0   1067.000000  
25%                 3.000000             0.0  30627.000000  
50%                 6.000000             0.0  49811.000000  
75%                 9.000000             0.0  72108.000000  
max                12.000000             0.0  99996.000000

In [20]:
autos['year_of_registration'].value_counts(normalize = True).head()

2000    0.069129
1999    0.064493
2005    0.063676
2001    0.059679
2006    0.059392
Name: year_of_registration, dtype: float64

According to the most frequent registration year, we can conclude that the german cars are younger than 20 years (the data were collected in 2016).

In [21]:
autos['brand'].value_counts(normalize= True).head(6)


volkswagen       0.206615
bmw              0.112029
opel             0.107458
mercedes_benz    0.095293
audi             0.090722
ford             0.069085
Name: brand, dtype: float64

German manufacturers make up the top 5 brands for the cars in the sales data with Volkswagen being the most popular, with approximately double the cars than the next two manufacturers combined.

We'll limit our analysis to brands that make up more than 5% of the total sales data.

In [22]:
# the most frequent cars
brands = ['volkswagen','bmw','opel','mercedes_benz','audi','ford']
# create boolean seris
boolean_freq_cars = []
for elem in autos['brand']:
    if elem in brands:
        boolean_freq_cars.append(True)
    else:
        boolean_freq_cars.append(False)
# use boolean list to filter the dataset
autos = autos[boolean_freq_cars]
autos.head()

date_crawled                                               name  \
0 2016-03-24 11:52:17                                         Golf_3_1.6   
1 2016-03-24 10:58:45                               A5_Sportback_2.7_Tdi   
3 2016-03-17 16:54:04                               GOLF_4_1_4__3T�RER   
5 2016-04-04 17:36:23  BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...   
8 2016-04-04 23:42:13               Ford_C___Max_Titanium_1_0_L_EcoBoost   

   seller offer_type  price_$   abtest vehicle_type  year_of_registration  \
0  privat    Angebot      480     test          NaN                  1993   
1  privat    Angebot    18300     test        coupe                  2011   
3  privat    Angebot     1500     test   kleinwagen                  2001   
5  privat    Angebot      650     test    limousine                  1995   
8  privat    Angebot    14500  control          bus                  2014   

   gearbox  power_ps  model  odometer_km  month_of_registration fuel_type  \
0  manuell         0   golf       150000                      0    benzin   
1  manuell       190    NaN       125000                      5    diesel   
3  manuell        75   golf       150000                      6    benzin   
5  manuell       102    3er       150000                     10    benzin   
8  manuell       125  c_max        30000                      8    benzin   

        brand not_repaired_damage date_created  nr_of_pictures  postal_code  \
0  volkswagen                 NaN   2016-03-24               0        70435   
1        audi                  ja   2016-03-24               0        66954   
3  volkswagen                nein   2016-03-17               0        91074   
5         bmw                  ja   2016-04-04               0        33775   
8        ford                 NaN   2016-04-04               0        94505   

            last_seen  
0 2016-04-07 03:16:57  
1 2016-04-07 01:46:50  
3 2016-03-17 17:40:17  
5 2016-04-06 19:17:07  
8 2016-04-04 23:42:13

In [23]:
autos_pivot = pd.pivot_table(autos, values=['price_$','odometer_km'], index=None, columns=['brand'], aggfunc='mean')
autos_pivot.head()

brand                 audi            bmw           ford  mercedes_benz  \
odometer_km  129453.638355  132790.697674  124562.160435  130781.974050   
price_$        9220.696763    8490.995270    3872.042506    8432.452502   

brand                 opel     volkswagen  
odometer_km  129090.815698  128536.546271  
price_$        2929.028560    7905.022120

We can see that the car mileage doesn't vary as much as the prices. We can see that the more expensive brands generally tend to have higher mileage with the only outlier being Opel.

Let's continue to clean up the data. To do this we'll indentify the categorical data that uses german words and translate and map the values to thir english counterparts

In [24]:
autos['gearbox'].value_counts()

manuell      22087
automatik     7477
Name: gearbox, dtype: int64

In [25]:
autos['not_repaired_damage'].value_counts()

nein    22631
ja       2891
Name: not_repaired_damage, dtype: int64

In [26]:
autos['fuel_type'].value_counts()

benzin     17372
diesel     10852
lpg          408
cng           59
andere         8
hybrid         5
elektro        4
Name: fuel_type, dtype: int64

In [27]:
autos['vehicle_type'].value_counts()

limousine     9808
kombi         6864
kleinwagen    5340
bus           2573
cabrio        1708
coupe         1702
suv            848
andere         214
Name: vehicle_type, dtype: int64

We should cretae the dictionaries with the old and new names, and use tham during the replace.

In [28]:
gearbox = {
    'manuell': 'manual',
    'automatik': 'automatic'
}

not_repaired_damage = {
    'nein': 'no',
    'ja': 'yes'
}

autos['gearbox'] = autos['gearbox'].map(gearbox)
autos['not_repaired_damage'] = autos['not_repaired_damage'].map(not_repaired_damage)
autos['fuel_type'] = autos['fuel_type'].replace(['benzin', 'elektro', 'andere'], ['gasoline', 'electric', 'other'])
autos['vehicle_type'] = autos['vehicle_type'].replace(['limousine', 'kleinwagen', 'kombi', 'cabrio', 'andere'], ['sedan', 'small car', 'stationwagen', 'convertible', 'other'])
autos['model'] = autos['model'].replace('andere', 'other')

autos.head()

date_crawled                                               name  \
0 2016-03-24 11:52:17                                         Golf_3_1.6   
1 2016-03-24 10:58:45                               A5_Sportback_2.7_Tdi   
3 2016-03-17 16:54:04                               GOLF_4_1_4__3T�RER   
5 2016-04-04 17:36:23  BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...   
8 2016-04-04 23:42:13               Ford_C___Max_Titanium_1_0_L_EcoBoost   

   seller offer_type  price_$   abtest vehicle_type  year_of_registration  \
0  privat    Angebot      480     test          NaN                  1993   
1  privat    Angebot    18300     test        coupe                  2011   
3  privat    Angebot     1500     test    small car                  2001   
5  privat    Angebot      650     test        sedan                  1995   
8  privat    Angebot    14500  control          bus                  2014   

  gearbox  power_ps  model  odometer_km  month_of_registration fuel_type  \
0  manual         0   golf       150000                      0  gasoline   
1  manual       190    NaN       125000                      5    diesel   
3  manual        75   golf       150000                      6  gasoline   
5  manual       102    3er       150000                     10  gasoline   
8  manual       125  c_max        30000                      8  gasoline   

        brand not_repaired_damage date_created  nr_of_pictures  postal_code  \
0  volkswagen                 NaN   2016-03-24               0        70435   
1        audi                 yes   2016-03-24               0        66954   
3  volkswagen                  no   2016-03-17               0        91074   
5         bmw                 yes   2016-04-04               0        33775   
8        ford                 NaN   2016-04-04               0        94505   

            last_seen  
0 2016-04-07 03:16:57  
1 2016-04-07 01:46:50  
3 2016-03-17 17:40:17  
5 2016-04-06 19:17:07  
8 2016-04-04 23:42:13

No let's investigate wether or not ther is a price discrepancy if cars that have been previously damaged compared to their don't damaged counterparts.

In [29]:
autos_pivot_damaged = pd.pivot_table(autos, values=['price_$'], index=None, columns=['not_repaired_damage'], aggfunc='mean')
print(autos_pivot_damaged)

not_repaired_damage           no          yes
price_$              7443.155406  2807.849533


We can see that cars that have been previously damaged are considerable cheaper than those with no damage prior to their listing.